### ISSUE TITLE
getNestedTuneResultsOptPathDf when using makeTuneControlMBO and benchmark

### ISSUE THREAD

**Issue by lucasxteixeira** [mlr-org/mlr#2725]
## Description

The function getNestedTuneResultsOptPathDf fails when using MBO as the optimization algorithm (makeTuneControlMBO) in a benchmark experiment (with keep.extract = TRUE).
The problem is that the opt.path is not directly in the extract slot, but inside the mbo.result.

In my case, I manually changed the benchmark object to fix that:
`bmr$results$data[[1]]$extract[[1]]$opt.path <- bmr$results$data[[1]]$extract[[1]]$mbo.result$opt.path`

**Comment by jakob-r** [comment]
It works for me:
```r
library(mlr)
library(mlrMBO)
lrn = makeTuneWrapper(makeLearner("classif.rpart"), resampling = makeResampleDesc("Holdout"), par.set = makeNumericParamSet("cp", 1, 0, 1), control = makeTuneControlMBO())
bmr = benchmark(lrn, sonar.task, makeResampleDesc("Holdout"), keep.extract = TRUE)
op = getNestedTuneResultsOptPathDf(bmr$results$`Sonar-example`$classif.rpart.tuned)
tail(op)
```

```
          cp mmce.test.mean   dob   eol error.message exec.time  iter
       <num>          <num> <int> <int>        <char>     <num> <int>
1: 0.1399174      0.3913043     9    NA          <NA>     0.017     1
2: 0.3508433      0.3913043    10    NA          <NA>     0.017     1
3: 0.2509032      0.3913043    11    NA          <NA>     0.018     1
4: 0.1574630      0.3913043    12    NA          <NA>     0.017     1
5: 0.3033800      0.3913043    13    NA          <NA>     0.018     1
6: 0.2014611      0.3913043    14    NA          <NA>     0.016     1
```

If you produce a minimal example that demonstrates that bug I will reopen this issue.

**Comment by lucasxteixeira** [comment]
I'm sorry for the lack of reproduce example.
The problem appears when I try to run it in parallel (with multiple proposed points):
```
library(mlr)
library(mlrMBO)
library(parallelMap)
mbo_ctrl <- mlrMBO::makeMBOControl(propose.points = 2)
mbo_ctrl <- mlrMBO::setMBOControlTermination(mbo_ctrl, iters = 2)
tune_ctrl <- mlr::makeTuneControlMBO(mbo.control = mbo_ctrl)
lrn = makeTuneWrapper(makeLearner("classif.rpart"), resampling = makeResampleDesc("Holdout"), par.set = makeNumericParamSet("cp", 1, 0, 1), control = tune_ctrl)
parallelStartMulticore(cpus = 2, level = "mlrMBO.feval", show.info = FALSE)
bmr = benchmark(lrn, sonar.task, makeResampleDesc("Holdout"), keep.extract = TRUE)
parallelStop()
ef = generateHyperParsEffectData(bmr$results$`Sonar-example`$classif.rpart.tuned)
ef
```
```
Error in as.data.frame.OptPathDF(X[[i]], ...) : 
  No elements where selected (via 'dob' and 'eol')!
```

**Comment by jakob-r** [comment]
What happens without parallelization:
mlr constructs an objective function for mlrMBO. In this function we have a pointer to the mlr-OptPath. Each call of the objective function will write into the mlr-OptPath.
On the other hand we have an mlrMBO-OptPath that gets build from within mlrMBO where we write the result of each call of the objective.

What happens in parallel:
Now the function evaluations happen in individual processes. Therefore the pointer does not point to the original mlr-OptPath (on the master) anymore but to a "local" copy. This local copy dies as soon as the individual process finishes. As we only return the result of the objective the changes to the mlr-OptPath are lost.
As a result the mlr-OptPath is empty after the optimization.

However, the mlrMBO-OptPath can be written normally because it happens on the "master" outside of the parallelization.
That's why you can still access the mlrMBO-OptPath `bmr$results$`Sonar-example`$classif.rpart.tuned$extract[[1]]$mbo.result$opt.path.

Long story short. This is a bit complicated and I would suggest to not fix it as mlr is in maintenance mode and we are developing mlr3 - hopefully without this problem.
Any opinions from the team? @berndbischl

**Comment by lucasxteixeira** [comment]
The $mbo.result$opt.path is not exactly the same as the original $opt.path. For instance, if a set multiple measures on the benchmark, they won't be in the $mbo.result$opt.path. So, a simple copy might not suffice in every situation.
Well, since this issue was never brought up, this configuration is probably rarely used.
Maybe just add an note in https://mlr.mlr-org.com/reference/makeTuneControlMBO.html to inform that using TuneControlMBO in parallel might break the workflow.
If you guys can point me some directions I can try to fix it.

**Comment by jakob-r** [comment]
Yes. They are two different OptPaths as I tried to indicate.
If you desperately need the mlr-OptPath and the parallelization you could clone mlr
change this line
https://github.com/mlr-org/mlr/blob/b12c250f9485fb9fff6f7bb88cd4309584f6de89/R/tunerFitnFun.R#L26
to something like 
```r
saveRDS(generate_free_filename(), list(x = x, y = res$y, dob = dob, eol = NA, check.feasible = TRUE, exec.time = res$exec.time, error.message = res$errmsg, extra = extra))
```
and after this line
https://github.com/mlr-org/mlr/blob/b57049d9ae4b73b8de7d012d648e7cca9a6478bc/R/tuneMBO.R#L21
you could iterate over all the stored OptPath elements and add them to the OptPath
```r
for (f in list_op_files()) {
  x = readRDS(f)
  do.call(addOptPathEl, c(list(opt.path = opt.path), x))
  unlink(f)
}
```

However, this is far from an elegant solution that could be applied generally because it relies on the file system to communicate from the parallel processes.

**Comment by pat-s** [comment]
Thanks Jakob for providing the workaround. "wonfixnow" is a good summary here as the issue is very niche and not easy to "fix".

Closing here for now, feel free to still comment if something is unclear.
